post_office/mail.py

```diff
@@ -268,12 +278,27 @@ def send(email):
 
     connections.close()
 
-    # Update statuses of sent and failed emails
+    # Update statuses of sent emails
     email_ids = [email.id for email in sent_emails]
     Email.objects.filter(id__in=email_ids).update(status=STATUS.sent)
 
-    email_ids = [email.id for (email, e) in failed_emails]
-    Email.objects.filter(id__in=email_ids).update(status=STATUS.failed)
+    # Conditionally update statuses of failed emails
+    email_ids = [email.id for email, _ in failed_emails]
+    emails_failed = Email.objects.filter(id__in=email_ids).annotate(
+        retry_counter=Case(
+            When(number_of_retries__isnull=True, then=Value(1)),
+            default=F('number_of_retries') + 1
+        )
+    )
+    query = Q(retry_counter__lte=get_max_retries())
+    num_failed = emails_failed.exclude(query).update(
+        status=STATUS.failed
+    )
+    num_requeued = emails_failed.filter(query).update(
+        number_of_retries=F('retry_counter'),
+        status=STATUS.requeued,
+        scheduled_time=timezone.now() + get_time_delta_to_retry(),
+    )
```

Comment:
I'm not a big fan of this code block because it makes two separate DB queries. Since Django 1.8 is EOL'ed, I think we should use `bulk_update()` (available in Django >= 2.2), the resulting code should also be more concise. Something like this:

```python
max_retries = get_max_retries()
for email in failed_emails:
    if email.number_of_retries < max_retries:
        email.status = STATUS.requeued
    # ...

Email.objects.bulk_update(failed_emails, ['status', 'number_of_retries']
```

Reply:
I changed that to a `bulk_update` logic.